pyrokinetics/numerics.py

```diff
@@ -1,58 +1,151 @@
-from cleverdict import CleverDict
-from copy import deepcopy
+import dataclasses
+import json
+import pprint
+from typing import Any, Dict, Optional
 
+from .metadata import metadata
 
-class Numerics(CleverDict):
-    """
-    Set up numerical grid
 
+@dataclasses.dataclass
+class Numerics:
+    """
+    Stores information describing numerical features common to most gyrokinetics
+    solvers, such as the dimensions of numerical grids, the presence of electromagnetic
+    field components, and time steps. Numerics is not used to store special flags
+    belonging to only one gyrokinetics code.
     """
 
-    def __init__(self, *args, **kwargs):
-        s_args = list(args)
-
-        if args and not isinstance(args[0], CleverDict) and isinstance(args[0], dict):
-            s_args[0] = sorted(args[0].items())
-
-        if args:
-            super().__init__(*s_args, **kwargs)
-        else:
-            self.default()
-
-    def default(self):
-        data_dict = {
-            "ntheta": 32,
-            "theta0": 0.0,
-            "nenergy": 8,
-            "npitch": 8,
-            "nky": 1,
-            "nkx": 1,
-            "kx": 0.0,
-            "ky": 0.1,
-            "nperiod": 1,
-            "nonlinear": False,
-            "phi": True,
-            "apar": False,
-            "bpar": False,
-            "max_time": 500.0,
-            "delta_time": 0.001,
-            "beta": None,
-        }
-
-        super().__init__(data_dict)
-
-    def __deepcopy__(self, memodict):
+    #: Number of elements in the :math:`\theta` (poloidal) grid
+    ntheta: int = 32
+
+    #: Number of :math:`2\pi` segments in the toroidal direction.
+    nperiod: int = 1
+
+    #: Number of elements in the energy grid
+    nenergy: int = 8
+
+    #: Number of elements in the pitch grid
+    npitch: int = 8
+
+    #: Number of elements in the velocity-space :math:`k_y` grid
+    nky: int = 1
+
+    #: Number of elements in the velocity-space :math:`k_x` grid
+    nkx: int = 1
+
+    #: Value of :math:`k_y\rho`
+    ky: float = 0.1
+
+    #: Value of :math:`k_x\rho`
+    kx: float = 0.0
+
+    #: Initial time step, in units of ``tref``
```

Comment:
Units of time step would be in ``lref / vref``

Reply:
Cheers, fixed now